File: clients/python/basemouse_client.py
"""Lightweight BaseMouse API client using only the Python standard library."""
from __future__ import annotations

import json
import os
import urllib.error
import urllib.parse
import urllib.request
from dataclasses import dataclass
from typing import Any, Mapping


class BaseMouseAPIError(RuntimeError):
    def __init__(self, message: str, *, status: int | None = None, body: Any = None, url: str | None = None):
        super().__init__(message)
        self.status = status
        self.body = body
        self.url = url


@dataclass
class BaseMouseClient:
    base_url: str = os.getenv("BASEMOUSE_URL", "https://basemouse.com")
    api_key: str = os.getenv("BASEMOUSE_API_KEY", os.getenv("BASEMOUSE_TOKEN", ""))
    timeout: float = 15.0

    def __post_init__(self) -> None:
        self.base_url = self.base_url.rstrip("/")

    def request(self, path: str, *, method: str = "GET", query: Mapping[str, Any] | None = None, body: Any = None) -> Any:
        url = self.base_url + path
        params = {k: str(v) for k, v in (query or {}).items() if v is not None and v != ""}
        if params:
            url += "?" + urllib.parse.urlencode(params)
        data = None if body is None else json.dumps(body).encode("utf-8")
        headers = {"Accept": "application/json"}
        if self.api_key:
            headers["Authorization"] = f"Bearer {self.api_key}"
        if data is not None:
            headers["Content-Type"] = "application/json"
        req = urllib.request.Request(url, data=data, method=method, headers=headers)
        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as res:  # nosec B310: caller controls BaseMouse URL intentionally
                text = res.read().decode("utf-8")
                return json.loads(text) if text else None
        except urllib.error.HTTPError as exc:
            raw = exc.read().decode("utf-8", errors="replace")
            try:
                parsed = json.loads(raw) if raw else None
            except json.JSONDecodeError:
                parsed = raw
            message = parsed.get("message") or parsed.get("error") if isinstance(parsed, dict) else str(parsed)
            raise BaseMouseAPIError(message or f"BaseMouse request failed with {exc.code}", status=exc.code, body=parsed, url=url) from exc
        except Exception as exc:  # network, timeout, JSON decode
            raise BaseMouseAPIError(str(exc), url=url) from exc

    def search(self, q: str, *, type: str | None = None, tag: str | None = None, retrieval: str | None = None, mode: str | None = None) -> Any:
        # `retrieval` picks the ranking mode ("lexical" default, "hybrid" for
        # graph-aware expansion); `mode` is accepted as an alias.
        return self.request("/api/search", query={"q": q, "type": type, "tag": tag, "retrieval": retrieval if retrieval is not None else mode})

    def context_pack(self, q: str | None = None, *, limit: int | None = None, type: str | None = None, tag: str | None = None, workspace: str | None = None, retrieval: str | None = None, mode: str | None = None) -> Any:
        return self.request("/api/context-pack", query={"q": q, "limit": limit, "type": type, "tag": tag, "workspace": workspace, "retrieval": retrieval if retrieval is not None else mode})

    def list_repository(self, *, limit: int | None = None, offset: int | None = None) -> Any:
        return self.request("/api/repository", query={"limit": limit, "offset": offset})

    def create_document(self, document: Mapping[str, Any]) -> Any:
        return self.request("/api/documents", method="POST", body=dict(document))

    def update_document(self, doc_id: str, fields: Mapping[str, Any], *, expected_version: int | None = None) -> Any:
        body = dict(fields)
        if expected_version is not None:
            body["expectedVersion"] = expected_version
        return self.request(f"/api/documents/{urllib.parse.quote(doc_id, safe='')}", method="PUT", body=body)

    def delete_document(self, doc_id: str) -> Any:
        return self.request(f"/api/documents/{urllib.parse.quote(doc_id, safe='')}", method="DELETE")

    def document_history(self, doc_id: str) -> Any:
        return self.request(f"/api/documents/{urllib.parse.quote(doc_id, safe='')}/history")

    def usage(self) -> Any:
        return self.request("/api/usage")

    def rotate_key(self) -> Any:
        return self.request("/api/keys/rotate", method="POST")


def format_context_pack_for_prompt(pack: Mapping[str, Any]) -> str:
    entries = pack.get("entries") or []
    if not entries:
        return "BaseMouse context: no matching entries."
    lines = ["BaseMouse context (cite labels when used):"]
    for entry in entries:
        citation = entry.get("citation") or {}
        provenance = entry.get("provenance") or {}
        relevance = entry.get("relevance") or {}
        label = citation.get("label") or f"[{entry.get('id')}] {entry.get('title')}"
        lines.append(f"\n{label}\nscore={relevance.get('score', 'n/a')}; checksum={provenance.get('checksum', 'n/a')}\n{entry.get('body', '')}")
    return "\n".join(lines)

File: clients/js/basemouse-client.js
/** Lightweight ESM BaseMouse API client. No runtime dependencies. */
export class BaseMouseAPIError extends Error {
  constructor(message, { status, body, url } = {}) {
    super(message);
    this.name = 'BaseMouseAPIError';
    this.status = status;
    this.body = body;
    this.url = url;
  }
}

export class BaseMouseClient {
  constructor({ baseUrl = process.env.BASEMOUSE_URL || 'https://basemouse.com', apiKey = process.env.BASEMOUSE_API_KEY || process.env.BASEMOUSE_TOKEN || '', fetchImpl = globalThis.fetch, timeoutMs = 15000 } = {}) {
    if (!fetchImpl) throw new Error('BaseMouseClient requires fetch (Node 18+ or a fetchImpl)');
    this.baseUrl = baseUrl.replace(/\/$/, '');
    this.apiKey = apiKey;
    this.fetchImpl = fetchImpl;
    this.timeoutMs = timeoutMs;
  }

  async request(path, { method = 'GET', query, body, headers = {} } = {}) {
    const url = new URL(path, this.baseUrl);
    for (const [key, value] of Object.entries(query || {})) {
      if (value !== undefined && value !== null && value !== '') url.searchParams.set(key, String(value));
    }
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), this.timeoutMs);
    const reqHeaders = { Accept: 'application/json', ...headers };
    if (this.apiKey) reqHeaders.Authorization = `Bearer ${this.apiKey}`;
    if (body !== undefined) reqHeaders['Content-Type'] = 'application/json';
    try {
      const res = await this.fetchImpl(url, {
        method,
        headers: reqHeaders,
        body: body === undefined ? undefined : JSON.stringify(body),
        signal: controller.signal
      });
      const text = await res.text();
      // Parse defensively: a proxy/load-balancer 502/504 returns an HTML body,
      // so parse failures must still surface the real HTTP status below rather
      // than masquerade as a generic "Unexpected token" error with no status.
      let parsed = null;
      if (text) {
        try { parsed = JSON.parse(text); } catch { parsed = text; }
      }
      if (!res.ok) {
        throw new BaseMouseAPIError(parsed?.message || parsed?.error || `BaseMouse request failed with ${res.status}`, { status: res.status, body: parsed, url: String(url) });
      }
      return parsed;
    } catch (error) {
      if (error instanceof BaseMouseAPIError) throw error;
      throw new BaseMouseAPIError(error.name === 'AbortError' ? 'BaseMouse request timed out' : error.message, { url: String(url) });
    } finally {
      clearTimeout(timeout);
    }
  }

  // `retrieval` selects the ranking mode ('lexical' default, or 'hybrid' for
  // graph-aware expansion). `mode` is accepted as an alias. Omitting both keeps
  // the back-compatible lexical behavior.
  search({ q, type, tag, retrieval, mode } = {}) {
    return this.request('/api/search', { query: { q, type, tag, retrieval: retrieval ?? mode } });
  }

  contextPack({ q, limit, type, tag, workspace, retrieval, mode } = {}) {
    return this.request('/api/context-pack', { query: { q, limit, type, tag, workspace, retrieval: retrieval ?? mode } });
  }

  listRepository({ limit, offset } = {}) {
    return this.request('/api/repository', { query: { limit, offset } });
  }

  createDocument(document) {
    return this.request('/api/documents', { method: 'POST', body: document });
  }

  updateDocument(id, fields, { expectedVersion } = {}) {
    const body = expectedVersion === undefined ? fields : { ...fields, expectedVersion };
    return this.request(`/api/documents/${encodeURIComponent(id)}`, { method: 'PUT', body });
  }

  deleteDocument(id) {
    return this.request(`/api/documents/${encodeURIComponent(id)}`, { method: 'DELETE' });
  }

  documentHistory(id) {
    return this.request(`/api/documents/${encodeURIComponent(id)}/history`);
  }

  usage() {
    return this.request('/api/usage');
  }

  rotateKey() {
    return this.request('/api/keys/rotate', { method: 'POST' });
  }
}

export function formatContextPackForPrompt(pack) {
  const entries = pack?.entries || [];
  if (!entries.length) return 'BaseMouse context: no matching entries.';
  const lines = ['BaseMouse context (cite labels when used):'];
  for (const entry of entries) {
    const label = entry.citation?.label || `[${entry.id}] ${entry.title}`;
    const score = entry.relevance?.score ?? 'n/a';
    const checksum = entry.provenance?.checksum || 'n/a';
    lines.push(`\n${label}\nscore=${score}; checksum=${checksum}\n${entry.body || ''}`);
  }
  return lines.join('\n');
}

export default BaseMouseClient;
